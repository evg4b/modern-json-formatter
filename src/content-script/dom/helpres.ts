import { ArrayNode, ObjectNode, TokenNode } from '@packages/tokenizer';
import { itemsCount, propertiesCount } from './elements';

export const element = (options?: { content?: string; class?: string; block?: boolean }) => {
  const span = document.createElement(!options?.block ? 'span' : 'div');
  if (options?.content) {
    span.appendChild(document.createTextNode(options.content));
  }
  if (options?.class) {
    span.className = options.class;
  }

  return span;
};

interface Options<K extends keyof HTMLElementTagNameMap = 'span'> {
  element: K;
  content?: string;
  class?: string | string[];
  children?: HTMLElement[] | HTMLElement;
  attributes?: { [key: string]: string };
}

export const createElement = <K extends keyof HTMLElementTagNameMap>(options: Options<K>): HTMLElementTagNameMap[K] => {
  const element = document.createElement(options.element ?? 'span');
  if (options.content) {
    element.appendChild(document.createTextNode(options.content));
  }
  if (options.class) {
    const classList = [options.class].flat();
    element.classList.add(...classList);
  }
  if (options.children) {
    const children = [options.children].flat();
    element.append(...children);
  }
  if (options.attributes) {
    Object.entries(options.attributes).forEach(([key, value]) => {
      element.setAttribute(key, value);
    });
  }

  return element;
};

export const isValueExpandable = (value: TokenNode): value is ObjectNode | ArrayNode =>
  (value.type === 'object' && !!value.properties.length) || (value.type === 'array' && !!value.items.length);

export const isToggleElement = (element: EventTarget | null): element is HTMLElement => {
  return element instanceof HTMLElement && element.classList.contains('toggle');
};

export const buildInfoNode = (value: TokenNode): HTMLSpanElement | null => {
  if (isValueExpandable(value)) {
    switch (value.type) {
      case 'array':
        return itemsCount(value.items.length);
      case 'object':
        return propertiesCount(value.properties.length);
      default:
        throw new Error('Unknown type');
    }
  }

  return null;
};
