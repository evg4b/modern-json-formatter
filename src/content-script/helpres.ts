export const throws = (value?: string): never => {
  throw new Error(value ?? 'Unexpected value');
};

// eslint-disable-next-line @typescript-eslint/no-unnecessary-type-parameters
export const query = <T extends HTMLElement = HTMLElement, D extends HTMLElement = HTMLElement>(
  value: D,
  selector: string
): T => value.querySelector<T>(selector) ?? throws(`Element ${selector} not found`);

const types: string[] = ['raw', 'query', 'formatted'] satisfies TabType[];

export function assetTabType(s?: string | null): asserts s is TabType {
  if (s && !types.includes(s)) {
    throw new Error(`Invalid tab type '${s}'`);
  }
}
