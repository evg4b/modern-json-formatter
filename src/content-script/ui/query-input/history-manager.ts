export class HistoryManager<T> {
  private readonly history: T[] = [];
  private historyIndex = -1;

  save(state: T) {
    if (this.historyIndex < this.history.length - 1) {
      this.history.splice(this.historyIndex + 1);
    }

    this.history.push(state);
    this.historyIndex = this.history.length - 1;
  }

  undo() {
    if (this.historyIndex > 0) {
      this.historyIndex--;
      return this.history[this.historyIndex];
    }
    return null;
  }

  redo() {
    if (this.historyIndex < this.history.length - 1) {
      this.historyIndex++;
      return this.history[this.historyIndex];
    }
    return null;
  }
}
