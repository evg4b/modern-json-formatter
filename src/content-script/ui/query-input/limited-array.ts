export class LimitedArray<T> extends Array<T> {
  constructor(private readonly limit: number) {
    super();
  }

  public push(...items: T[]): number {
    if (this.length >= this.limit) {
      this.shift();
    }
    return super.push(...items);
  }
}
