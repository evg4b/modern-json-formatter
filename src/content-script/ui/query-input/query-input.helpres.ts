export const isUndoEvent = (event: KeyboardEvent) => {
  return (event.metaKey || event.ctrlKey) && event.key === 'z' && !event.shiftKey;
};

export const isRedoEvent = (event: KeyboardEvent) => {
  return (event.metaKey || event.ctrlKey) && event.key === 'z' && event.shiftKey;
};

export const isWrapEvent = (event: KeyboardEvent, brackets: Record<string, string>) => {
  return event.key in brackets;
};

export const isSubmitEvent = (event: KeyboardEvent) => {
  return event.key === 'Enter';
};

export const isPrintableKey = (event: KeyboardEvent) => {
  return event.key.length === 1
    && !event.shiftKey
    && !event.ctrlKey
    && !event.metaKey
    && !event.altKey;
}
