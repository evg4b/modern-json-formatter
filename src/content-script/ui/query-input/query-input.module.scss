:host {
  display: flex;
  flex-direction: row;
  gap: 5px;

  input {
    min-width: 200px;
    width: 30vw;
  }

  .input-wrapper {
    display: flex;
    flex-direction: column;
    gap: 3px;
    position: relative;

    .error-message {
      position: absolute;
      top: calc(100% + 5px);
      color: red;
      font-size: 10px;
      user-select: none;

      background: var(--background-color);
      padding: 2px;
      box-shadow: 0 0 2px 1px var(--background-color);

      &.hidden {
        display: none;
      }
    }
  }
}
