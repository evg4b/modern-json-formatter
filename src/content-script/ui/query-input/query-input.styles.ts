export default `
:host {
  display: flex;
  flex-direction: row;
  gap: 5px;

  input {
    min-width: 200px;
    width: 30vw;
    background: #3b3b3b;
    border: 1px solid #858585;
    border-radius: 2px;
  }

  .input-wrapper {
    display: flex;
    flex-direction: column;
    gap: 3px;
    position: relative;

    input {
      flex: 1 1 auto;
    }

    .error-message {
      position: absolute;
      top: calc(100% + 5px);
      color: #e46962;
      font-size: 10px;
      user-select: none;

      background: var(--background-color);
      padding: 2px;
      box-shadow: 0 0 2px 1px var(--background-color);

      &.hidden {
        display: none;
      }
    }
  }
}
`
