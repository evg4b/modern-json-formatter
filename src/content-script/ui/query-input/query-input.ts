import { getHistory, pushHistory } from '@core/background';
import { getURL } from '@core/browser';
import { createElement, CustomElement, StyledComponentElement } from '@core/dom';
import { debounce } from 'lodash';
import { isNotNil } from '../../helpers';
import { InfoButtonElement } from '../info-button';
import { QueryHistory } from '../query-history/query-history';
import { HistoryManager } from './history-manager';
import { isRedoEvent, isSubmitEvent, isUndoEvent, isWrapEvent } from './query-input.helpers';
import queryInputStyles from './query-input.module.scss';

export const brackets: Record<string, string> = {
  '[': ']',
  '(': ')',
  '{': '}',
  '\'': '\'',
  '"': '"',
  '`': '`',
};

interface HistoryItem {
  query: string;
  start: number | null;
  end: number | null;
}

@CustomElement('query-input')
export class QueryInputElement extends StyledComponentElement {
  private readonly inputElement = this.createInput();
  private readonly errorMessageElement = createElement({
    element: 'span',
    class: ['error-message', 'hidden'],
  });
  private readonly queryHistoryElement = new QueryHistory();
  private readonly wrapperElement = createElement({
    element: 'div',
    class: 'input-wrapper',
    children: [
      this.inputElement,
      this.errorMessageElement,
      this.queryHistoryElement,
    ],
  });

  private readonly infoIcons = new InfoButtonElement(getURL('faq.html'));
  private readonly history = new HistoryManager<HistoryItem>();

  private onSubmitCallback: ((s: string) => unknown) | null = null;

  constructor() {
    super(queryInputStyles);
    this.shadow.append(this.infoIcons, this.wrapperElement);
    this.setupEventHandlers(this.inputElement);
    this.saveState();
    this.queryHistoryElement.onSelectedQuery(query => {
      this.inputElement.value = query;
      this.onSubmitEvent();
    });
  }

  public setErrorMessage(errorMessage: string | null): void {
    if (errorMessage) {
      this.errorMessageElement.textContent = errorMessage;
      this.errorMessageElement.classList.remove('hidden');
    } else {
      this.errorMessageElement.textContent = '';
      this.errorMessageElement.classList.add('hidden');
    }
  }

  public onSubmit(callback: (s: string) => void | Promise<void>): void {
    this.onSubmitCallback = callback;
  }

  public focus(): void {
    this.inputElement.focus();
  }

  public blur(): void {
    this.inputElement.blur();
  }

  public hide(): void {
    this.style.display = 'none';
  }

  public show(): void {
    this.style.display = 'flex';
  }

  private createInput(): HTMLInputElement {
    const input = createElement({ element: 'input' });
    input.type = 'text';
    input.placeholder = 'Input jq query...';

    return input;
  }

  private setupEventHandlers(input: HTMLInputElement) {
    input.addEventListener('keydown', event => {
      this.setErrorMessage(null);
      if (isSubmitEvent(event)) {
        this.onSubmitEvent();
      } else if (isWrapEvent(event, brackets)) {
        this.onWrapEvent(event);
      } else if (isUndoEvent(event)) {
        this.onUndoEvent(event);
      } else if (isRedoEvent(event)) {
        this.onRedoEvent(event);
      }
    });

    input.addEventListener('input', () => {
      this.saveState();
      void this.loadHistory(this.inputElement.value);
      console.log('input', this.inputElement.value);
    });

    input.addEventListener('focus', () => {
      this.queryHistoryElement.open();
    });
  }

  private onSubmitEvent() {
    this.onSubmitCallback?.(this.inputElement.value);
    this.queryHistoryElement.close();
    pushHistory(window.location.hostname, this.inputElement.value);
  }

  private readonly loadHistory = debounce(async (prefix: string) => {
    const history = await getHistory(window.location.hostname, prefix);
    this.queryHistoryElement.setHistory(history);
    this.queryHistoryElement.open();
  }, 300);

  private onWrapEvent(event: KeyboardEvent) {
    const start = this.inputElement.selectionStart ?? 0;
    const end = this.inputElement.selectionEnd ?? 0;
    if (isNotNil(start) && isNotNil(end) && start !== end) {
      event.preventDefault();
      const selectedText = this.inputElement.value.substring(start, end);
      this.inputElement.setRangeText(`${ event.key }${ selectedText }${ brackets[event.key] }`);
      this.inputElement.setSelectionRange(start + 1, end + 1);
      this.saveState();
    }
  }

  private onUndoEvent(event: KeyboardEvent) {
    event.preventDefault();
    const previousState = this.history.undo();
    this.restoreState(previousState);
  }

  private onRedoEvent(event: KeyboardEvent) {
    event.preventDefault();
    const nextState = this.history.redo();
    this.restoreState(nextState);
  }

  private saveState() {
    this.history.save({
      query: this.inputElement.value,
      start: this.inputElement.selectionStart,
      end: this.inputElement.selectionEnd,
    });
  }

  private restoreState(state: HistoryItem | null) {
    if (state) {
      this.inputElement.value = state.query;
      this.inputElement.setSelectionRange(state.start, state.end);
    }
  }
}
