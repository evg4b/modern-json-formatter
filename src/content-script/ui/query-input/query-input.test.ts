/* eslint-disable @typescript-eslint/unbound-method */
import '@testing/browser.mock';
import { beforeEach, describe, expect, test } from '@jest/globals';
import { getShadowRoot } from '@testing/styled-component';
import { throws } from '../../helpres';
import { brackets, QueryInputElement } from './query-input';
import { isRedoEvent, isSubmitEvent, isUndoEvent, isWrapEvent } from './query-input.helpres';

describe('QueryInputElement', () => {
  let input: QueryInputElement;
  let innerInput: HTMLInputElement;
  let shadowRoot: ShadowRoot;

  const keyPress = (key: string, options?: KeyboardEventInit) => {
    const fakeEvent = new KeyboardEvent('keydown', { key, ...(options ?? {}) });
    innerInput.dispatchEvent(fakeEvent);
    if (
      !isRedoEvent(fakeEvent)
      && !isUndoEvent(fakeEvent)
      && !isSubmitEvent(fakeEvent)
      && !isWrapEvent(fakeEvent, brackets)
    ) {
      innerInput.value += key;
      innerInput.dispatchEvent(new Event('input', { bubbles: true }));
    }
  };

  const undo = () => {
    keyPress('z', { ctrlKey: true });
  };

  const redo = () => {
    keyPress('z', { ctrlKey: true, shiftKey: true });
  };

  const times = (n: number, fn: () => void) => {
    Array.from({ length: n })
      .forEach(fn);
  };

  beforeEach(() => {
    input = new QueryInputElement();
    shadowRoot = getShadowRoot(input);
    innerInput = shadowRoot.querySelector('input') ?? throws('No input');
  });

  test('should create input element', () => {
    expect(innerInput).not.toBeNull();
  });

  describe('errors', () => {
    let errorMessage: HTMLElement | null;

    beforeEach(() => {
      errorMessage = shadowRoot.querySelector('.error-message');
    });

    test('should exists', () => {
      expect(errorMessage).not.toBeNull();
    });

    test('should have no error message by defaults', () => {
      input.setErrorMessage(null);

      expect(errorMessage?.classList.contains('hidden')).toBe(true);
      expect(errorMessage?.textContent).toEqual('');
    });

    test('should show error message', () => {
      input.setErrorMessage('Error message');

      expect(errorMessage?.classList.contains('hidden')).toBe(false);
      expect(errorMessage?.textContent).toEqual('Error message');
    });

    test('should hide error message', () => {
      input.setErrorMessage('Error message');
      input.setErrorMessage(null);

      expect(errorMessage?.classList.contains('hidden')).toBe(true);
      expect(errorMessage?.textContent).toEqual('');
    });
  });

  describe('show/hide', () => {
    test('should visible by default', () => {
      expect(input.style.display).not.toEqual('none');
    });

    test('should hide', () => {
      input.hide();

      expect(input.style.display).toEqual('none');
    });

    test('should show', () => {
      input.hide();
      input.show();

      expect(input.style.display).not.toEqual('none');
    });
  });

  describe('focus/blur', () => {
    test('should focus', () => {
      Reflect.set(innerInput, 'focus', jest.fn());

      input.focus();

      expect(innerInput.focus).toBeCalled();
    });

    test('should blur', () => {
      Reflect.set(innerInput, 'blur', jest.fn());

      input.blur();

      expect(innerInput.blur).toBeCalled();
    });
  });

  describe('typing', () => {
    test('should call onSubmitCallback on Enter', () => {
      const onSubmitCallback = jest.fn();
      input.onSubmit(onSubmitCallback);
      keyPress('Enter');

      expect(onSubmitCallback).toBeCalled();
    });

    test('should not call onSubmitCallback on other key', () => {
      const onSubmitCallback = jest.fn();
      input.onSubmit(onSubmitCallback);
      keyPress('a');

      expect(onSubmitCallback).not.toBeCalled();
    });

    test('should clear error message on typing', () => {
      input.setErrorMessage('Error message');
      keyPress('a');

      expect(shadowRoot.querySelector('.error-message')?.classList.contains('hidden')).toBe(true);
    });
  });

  describe('for selection', () => {
    const query = '.[] | key, value';
    const expectedSelection = 'key, value';

    beforeEach(() => {
      innerInput.value = '.[] | key, value';
      innerInput.focus();
    });

    const items = [
      { key: '[', expected: '.[] | [key, value]' },
      { key: '(', expected: '.[] | (key, value)' },
      { key: '{', expected: '.[] | {key, value}' },
      { key: '\'', expected: '.[] | \'key, value\'' },
      { key: '"', expected: '.[] | "key, value"' },
      { key: '`', expected: '.[] | `key, value`' },
    ];

    describe('when text is selected', () => {
      const start = query.indexOf(expectedSelection);
      const end = start + expectedSelection.length;

      beforeEach(() => {
        innerInput.setSelectionRange(start, end);
      });

      describe.each(items)('when $key is pressed', ({ key, expected }) => {
        beforeEach(() => {
          innerInput.dispatchEvent(new KeyboardEvent('keydown', { key }));
        });

        test('should save selection', () => {
          const selectedText = innerInput.value.substring(innerInput.selectionStart ?? 0, innerInput.selectionEnd ?? 0);
          expect(selectedText).toEqual(expectedSelection);
        });

        test(`should wrap selected text with ${ key }`, () => {
          expect(innerInput.value).toEqual(expected);
        });
      });
    });

    describe('for selected text', () => {
      beforeEach(() => {
        innerInput.setSelectionRange(query.length, query.length);
      });

      describe.each(items)('when $key is pressed', ({ key }) => {
        beforeEach(() => {
          innerInput.dispatchEvent(new KeyboardEvent('keydown', { key }));
        });

        test(`should not change text`, () => {
          expect(innerInput.value).toEqual(query);
        });
      });
    });
  });

  describe('history', () => {
    const select = (start: number, end: number) => {
      innerInput.setSelectionRange(start, end);
    };

    beforeEach(() => {
      keyPress('1');
      select(0, 1);
      keyPress('2');
      select(1, 2);
      keyPress('3');
      select(2, 3);
      keyPress('4');
      select(2, 4);
      keyPress('5');
    });

    describe('undo', () => {
      test('should undo value', () => {
        undo();

        expect(innerInput.value).toEqual('1234');
      });

      test('should undo value multiple times', () => {
        times(3, () => undo());

        expect(innerInput.value).toEqual('12');
      });

      test('should not undo if no more states', () => {
        times(10, () => undo());

        expect(innerInput.value).toEqual('');
      });
    });

    describe('redo', () => {
      beforeEach(() => {
        times(3, () => undo());
      });

      test('should redo value', () => {
        redo();
        expect(innerInput.value).toEqual('123');
      });

      test('should redo multiple times', () => {
        times(2, () => redo());

        expect(innerInput.value).toEqual('1234');
      });

      test('should not redo if no more states', () => {
        times(10, () => redo());

        expect(innerInput.value).toEqual('12345');
      });

      test('should not redo after changing', () => {
        times(3, () => keyPress('#'));
        redo();

        expect(innerInput.value).toEqual('12###');
      });
    });
  });
});
