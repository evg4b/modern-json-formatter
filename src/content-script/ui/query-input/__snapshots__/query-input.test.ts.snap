// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`QueryInputElement query history should load history on focus 1`] = `
<datalist
  id="history-list"
>
  <option
    value=".[]"
  >
    .[]
  </option>
  <option
    value=".[0]"
  >
    .[0]
  </option>
</datalist>
`;
